Set cwd before executing a project

Akirathan opened the pull request:
Fixes #12301

### Pull Request Description

Set working directory to the parent of PROJECT_ROOT before executing project. This ensures that all the relative paths are resolved as expected. This only affects the `engine-runner`, i.e., the cmdline execution.

### Important Notes

Implemented via native code in [WorkingDirectory](https://github.com/enso-org/enso/blob/59f7328965af29aa66a5e386de7476411cf5edb1/lib/java/os-environment/src/main/java/org/enso/os/environment/chdir/WorkingDirectory.java). The wdir change is done as soon as possible from [Main.launch](https://github.com/enso-org/enso/blob/59f7328965af29aa66a5e386de7476411cf5edb1/engine/runner/src/main/java/org/enso/runner/Main.java#L1454-L1456). See the explanation in [this comment](https://github.com/enso-org/enso/pull/12618#issuecomment-2778451448) why it needs to be done before `user.dir` system property is accessed.

Implemented in new [os-environment project](https://github.com/enso-org/enso/blob/6b4e49d93bad18659e3c116a8e8db382cc730adf/build.sbt#L4181) that sets up native image for running jUnit tests. `desktop-environment` was merged into this project, so that it is tested in NI as well.

Reverted relevant parts of #10660 and #10928 - see https://github.com/enso-org/enso/pull/12618#issuecomment-2757589066

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
Add failing Runner_Spec test

Commit:
Update docs

Commit:
Move Platform from desktop-environment to engine-common

Commit:
Move Platform from desktop-environment to engine-common

Commit:
Fix compilation of project-manager

Commit:
Fix compilation of project-manager

Commit:
Change working dir in Context.Builder

Commit:
Revert "Fix compilation of project-manager"

This reverts commit 8b7479367c087b24ac27132fb0ae2b844a9db2a8.

Commit:
Revert "Move Platform from desktop-environment to engine-common"

This reverts commit a923150a00b4d88554d864c21ef5310a50f09482.

Commit:
Change working dir if parent of project root exists

Commit:
Runner_Spec does not create any files.

And does not use java.nio.Files

Akirathan commented:
Setting current working directory via `System.setProperty("user.dir", path)` [does not work](https://github.com/enso-org/enso/issues/12301#issuecomment-2754029168). Let's try to set the current directory via polyglot context builder - https://github.com/enso-org/enso/pull/12618/commits/020c7358930366529107f447e5f5439941b1fcb7.

Commit:
Move setting working dir functionality from COntextFactory to EnsoLanguage

Commit:
Remove Runner_Spec and replace it with test in runtinme-integration-tests.

Runner_Spec was failing on NI anyway. There is no way of getting path to the currently running binary.

Commit:
Do not explicitly set working dir in build_tools.

This reverts part of #10660.

Commit:
Resolve relative paths test in File_Spec is not ignored.

This reverts part of #10660

Commit:
Revert #10928

Akirathan commented:
Relevant parts from #10660 reverted in:
- e5c35f28e903d2a2952c0d8a050471fbf3c95bca
- 4032f8acaa05d0f7524bd2b6076838eeaed6f6b6

#10928 reverted in c0df4388ddddba519dcf0216e27bf90c5d082078

Akirathan commented:
Tested:
- `runEngineDistribution --run /home/pavel/tmp/Proj/src/Main.enso`
- `runEngineDistribution --run test/Base_Tests/src/Data/Vector_Spec.enso`

Both works well without any warnings.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/EnsoLanguage.java:
```diff
@@ -166,6 +167,18 @@ protected EnsoContext createContext(Env env) {
+        var parentDirAbs = parentDir.toAbsolutePath().toString();
+        if (!System.getProperty("user.dir").equals(parentDirAbs)) {
+          var truffleFile = env.getPublicTruffleFile(parentDirAbs);
+          env.setCurrentWorkingDirectory(truffleFile);
```
The [javadoc](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleLanguage.Env.html#setCurrentWorkingDirectory(com.oracle.truffle.api.TruffleFile)) says:

> The current working directory is used to resolve non absolute paths in [TruffleFile](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleFile.html) methods.

Is this enough? How does:
```
polyglot java import java.io.File
main =
            file = File.new "MY_FILE.txt"
            file.absolute.path
```
behave?

Akirathan replied:
Unfortunately you are right. The test with `java.io.File`, added in b6ae53402be579f3573d02f52fd20a26e19c8b6a, fails:
```
[error] Test org.enso.interpreter.test.ContextInitTest.shouldChangeWorkingDir_BeforeExecutingProject_JavaFile failed: java.lang.AssertionError: 
[error] Expected: is "/tmp/junit14724631386455624624/MY_FILE.txt"
[error]      but: was "/home/pavel/dev/enso/engine/runtime-integration-tests/MY_FILE.txt", took 4.496 sec
[error]     at org.enso.interpreter.test.ContextInitTest.lambda$shouldChangeWorkingDir_BeforeExecutingProject_JavaFile$1(ContextInitTest.java:64)
[error]     at org.enso.test.utils.ProjectUtils.testProjectRun(ProjectUtils.java:108)
[error]     at org.enso.interpreter.test.ContextInitTest.shouldChangeWorkingDir_BeforeExecutingProject_JavaFile(ContextInitTest.java:59)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
[info] Test run org.enso.interpreter.test.ContextInitTest finished: 1 failed, 0 ignored, 2 total, 6.832s
[error] Failed: Total 2, Failed 1, Errors 0, Passed 1
[error] Failed tests:
[error] 	org.enso.interpreter.test.ContextInitTest
[error] (Test / testOnly) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 113 s (01:53), completed Mar 27, 2025, 6:22:45 PM
```

I will need to implement this via JNI and System Java after all ...

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ContextInitTest.java:
```diff
@@ -0,0 +1,41 @@
+    var projDir = tmpFolder.newFolder().toPath();
+    var mainSrc =
+        """
+        from Standard.Base import all
```
Please duplicate this test to directly use `polyglot import java.io.File`.

Akirathan replied:
Duplicated in b6ae53402be579f3573d02f52fd20a26e19c8b6a

JaroslavTulach reviewed (commented):
- trying to avoid usage of JNI and System Java?
- I am skeptical we can get this done without it
- Please expand the test cases to use `java.io.File` via host interop

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ContextInitTest.java:
```diff
@@ -0,0 +1,41 @@
+    var ctxBuilder = ContextUtils.defaultContextBuilder().out(out).err(out);
+    var expectedWorkingDir = projDir.getParent();
+    var expectedPath = expectedWorkingDir.resolve("MY_FILE.txt").toAbsolutePath().toString();
+    ProjectUtils.testProjectRun(
```
Better to turn this (also) into integration test that launches `enso` with and without `--jvm` flag. Such tests are traditionally placed into `Example_Tests` - like https://github.com/enso-org/enso/blob/ed63362051842cdff6391423eeb6f33d3371df87/test/Examples_Tests/src/Python_Examples_Spec.enso#L91

Akirathan replied:
As noted in the commit message of https://github.com/enso-org/enso/pull/12618/commits/ae2b06c18a711a50b350176e509df1721f22a898, I have already tried to test it that way, but the test was failing in NI, as `enso_bin` could not be found. If you could think about a way how to launch something like `enso --new New_Project` from within enso running in NI, I will reintroduce that test once again.

JaroslavTulach replied:
> There is no way of getting path to the currently running binary

There is a way. I was working with @sdedic to get such functionality into NI few years ago...

> how to launch something like enso --new New_Project from within enso running in NI

..one can `getClass().getProtectioDomain().getCodeSource().getLocation()` to get _"some URL"_ - probably the URL should point to the **location of `enso` executable**. Invoking that from a `.enso` test may need further changes:
- either some builtin 
- or reflective config

Akirathan replied:
At the end, there already was a test in [File_Spec](https://github.com/enso-org/enso/blob/2af704dd383ee89e19baf13db1c0c365ff5d350a/test/Base_Tests/src/System/File_Spec.enso#L377) that I re-enabled, and now is only ignored in JVM mode.

Commit:
Duplicate test with java.io.File

Commit:
Do not attempt to change dir via Language.Env.currentDirectory.

This is not sufficient. See https://github.com/enso-org/enso/pull/12618#discussion_r2017230447.

JaroslavTulach commented on engine/common/src/main/java/org/enso/common/Platform.java:
```diff
@@ -13,19 +12,9 @@ public enum Platform {
   private Platform() {}
 
-  private static Platform detectOperatingSystem() {
+  public static Platform detectOperatingSystem() {
```
We probably want to make this a _compilation constant_ during NI build time, so I suggest:
```suggestion
  private static Platform detectOperatingSystem() {
```
and rather make `OPERATING_SYSTEM` publicly visible. Possibly rename the field. Maybe `CURRENT`?

Akirathan replied:
Done in 62c9b63e690c33e64bb432a7e49953de4f2fefd4

JaroslavTulach commented on lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/Directories.java:
```diff
@@ -2,9 +2,17 @@
 
 /** Provides information about user directories. */
 public sealed interface Directories permits LinuxDirectories, MacOsDirectories, WindowsDirectories {
+  static Directories getForCurrentPlatform() {
```
Here we may also benefit from `CURRENT` being a compilation constant during NI build time. Make it a field:
```suggestion
  public static final Directories CURRENT = switch (...) {}
```

Akirathan replied:
Done in 48ae55b77fcfed3d289b299b4eb7e49590337f41

JaroslavTulach commented on lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/LinuxDirectories.java:
```diff
@@ -6,6 +6,16 @@
+
+  private LinuxDirectories() {}
+
+  static LinuxDirectories getInstance() {
```
Lazy initialization makes no sense. Either:
```
private static final LinuxDirectories INSTANCE = new LinuxDirectories()
```
or just `new LinuxDirectories()` in the [CURRENT = switch](https://github.com/enso-org/enso/pull/12618/files#r2018098336).

Akirathan replied:
Done in a99333153fded4a2ef03498d85bdb9c05fb6fca4

JaroslavTulach commented on lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/LinuxTrashBin.java:
```diff
@@ -22,13 +22,24 @@
   private static final String PATH_INFO = "info";
 
-  private final LinuxDirectories directories = new LinuxDirectories();
+  public static LinuxTrashBin getInstance() {
```
At least _encapsulate_:
```suggestion
  static LinuxTrashBin getInstance() {
```
but probably do way more. I really don't think this _lazy initialization pattern_ is good for anything!

Akirathan replied:
Done in a99333153fded4a2ef03498d85bdb9c05fb6fca4 and 6951fd305322e86d009af82181ca101051fb3cf3

JaroslavTulach commented on lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/MacOsDirectories.java:
```diff
@@ -4,6 +4,16 @@
 import java.nio.file.Path;
 
 final class MacOsDirectories implements Directories {
+  private static MacOsDirectories instance;
```
Just use `new MacOSDirectories()` in https://github.com/enso-org/enso/pull/12618/files#r2018098336

Akirathan replied:
Done in b30fa1b8703bcb6ecbdf917e6ad8a2b372b2907f

JaroslavTulach commented on lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/TrashBin.java:
```diff
@@ -1,9 +1,17 @@
 
 /** Operations with system trash */
 public sealed interface TrashBin permits LinuxTrashBin, WindowsTrashBin, MacTrashBin {
+  static TrashBin getForCurrentPlatform() {
```
Make `public static final` field.

Akirathan commented:
I attempted to implement and test working directory changing in repo https://github.com/Akirathan/chdir-native. Linux part is OK, but Windows tests on GH hosted runners were still failing. Last fail is in https://github.com/Akirathan/chdir-native/actions/runs/14173427574/job/39702423890#step:6:1876. Failing in linker. I think this is an issue with how I set up the Windows environment in those jobs. I will continue the work here.

<blockquote><img src="https://opengraph.githubassets.com/6f3866e35de2a7f68727c1f265b85a06476150196660fab831cfc56e85ada61f/Akirathan/chdir-native" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/Akirathan/chdir-native">GitHub - Akirathan/chdir-native: Does java.io in NI respect cwd?</a></strong></div><div>Does java.io in NI respect cwd? Contribute to Akirathan/chdir-native development by creating an account on GitHub.</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/6f3866e35de2a7f68727c1f265b85a06476150196660fab831cfc56e85ada61f/Akirathan/chdir-native" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/Akirathan/chdir-native">attempt to fix linking error on windows during NI build - shell32 · Akirathan/chdir-native@c1e6107</a></strong></div><div>Does java.io in NI respect cwd? Contribute to Akirathan/chdir-native development by creating an account on GitHub.</div></blockquote>

Commit:
Create separate `change-directory` project.

With NativeImage tests.

Commit:
Use TestCollectorFeature instead of manually created reflect-config.json

Akirathan commented:
On https://github.com/enso-org/enso/pull/12618/commits/fe862e1ff05a695bf78ef8850f3fc873c11c7b0f, we can run `sbt:change-directory/test` which builds NativeImage and runs [TestRunner](https://github.com/enso-org/enso/blob/94372e64cb11a8bf08239807b468db8a95dd4a09/lib/java/change-directory/src/test/java/org/enso/change/directory/TestRunner.java) in native image. The native image is lightweight, the build on my machine takes just 18 seconds. If we want to have some unit test for this feature, this is the easiest and proper way to do that - refactor it into a separate project, and build test NativeImage.

The test passes for Linux at the moment, I will add more platforms.

In the future, we can implement our own `junit-platform-native` inspired by [graalvm/native-build-tools/junit-platform-native](https://github.com/graalvm/native-build-tools/blob/master/common/junit-platform-native/README.md).

JaroslavTulach commented on build.sbt:
```diff
@@ -4139,6 +4141,62 @@ lazy val `benchmarks-common` =
+          "test-change-directory",
+          staticOnLinux = true,
+          targetDir     = targetDir,
+          mainClass     = Some("org.enso.change.directory.TestRunner"),
```
An effort to run unit tests in _native image_. Interesting.
- there is JUnit support: https://github.com/graalvm/native-build-tools/blob/master/common/junit-platform-native/README.md
- also built into Maven https://graalvm.github.io/native-build-tools/0.9.13/maven-plugin.html#testing-support

How does your work, @Akirathan, relates to these existing efforts?

Akirathan replied:
jUnit support in `graalvm/native-build-tools` works as a plugin for Gradle and Maven. My first idea was to almost copy & paste (of course with proper notices and license checks) the source code in https://github.com/graalvm/native-build-tools/tree/master/common/junit-platform-native into yet another project. But I abandoned that idea as too overengineered and just implemented [TestCollectorFeature](https://github.com/enso-org/enso/blob/fe862e1ff05a695bf78ef8850f3fc873c11c7b0f/lib/java/change-directory/src/test/java/org/enso/change/directory/TestCollectorFeature.java) and [TestRunner](https://github.com/enso-org/enso/blob/94372e64cb11a8bf08239807b468db8a95dd4a09/lib/java/change-directory/src/test/java/org/enso/change/directory/TestRunner.java) that are not based on the aforementioned `native-build-tools` at all. It is just a simple way how to run a list of jUnit classes (test suites) that is explicitly enumerated inside native image.

Akirathan replied:
Another reason for abandoning that idea is that we don't yet have a use case for running whole unit tests in native image. Once we have that, we can think about implementing a proper support based on https://github.com/graalvm/native-build-tools/tree/master/common/junit-platform-native.

Akirathan replied:
Of course, the reason **for** having unit tests in native image in this case is that the directory changing functionality **cannot be tested without native image**. And, same as you, I believe that *everything* should be tested, as long as the effort to create the tests does not exceed some reasonable limit.

JaroslavTulach replied:
If you know the alternatives, you know what you are doing. Yes, keeping it simple is good enough for CWD. Good luck.

Commit:
detectOperating system is private

Commit:
TrashBin and Directories have CURRENT static final fields

Commit:
No lazy initialization of static INSTANCE

Commit:
Decrease visibility

Commit:
No lazy initialization of static INSTANCE

Commit:
LinuxWorkingDirectory is singleton

Commit:
Test change dir

Commit:
Implement WorkingDirectory.exists

Commit:
fmt

Commit:
Fail `change-directory/test` task if some of the tests failed

Commit:
Add WindowsWorkingDirectory

JaroslavTulach commented on lib/java/change-directory/src/main/java/org/enso/change/directory/WorkingDirectory.java:
```diff
@@ -0,0 +1,19 @@
+  static WorkingDirectory getForCurrentPlatform() {
+    return switch (Platform.getOperatingSystem()) {
+      case Platform.LINUX -> LinuxWorkingDirectory.INSTANCE;
+      case Platform.MACOS -> throw new UnsupportedOperationException("unimplemented");
```
Mac OS X is Unix. That's going to be 1:1 to Linux. The CWD functionality is POSIX specified - e.g. same across all the Unixes.

JaroslavTulach commented on lib/java/change-directory/src/test/java/org/enso/change/directory/TestChangeDirectory.java:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  @Test
+  public void curDir() {
```
The order of methods as executed by JUnit doesn't follow the order in the source code. Make sure the tests are "idempotent"  and don't influence each other.

Details: https://github.com/junit-team/junit4/pull/293

Commit:
Remove outdated test.

Test outdated since 04242a0d4a1cf64ea1f4942e4fcf86cd254e797f. Will be handled by tests in change-directory project.

Commit:
Remove static fields from interfaces

Commit:
fmt

Akirathan commented:
The [linker error](https://github.com/enso-org/enso/actions/runs/14193180450/job/39762380040?pr=12618#step:10:1542) problem on Windows was resolved in https://github.com/Akirathan/chdir-native/commit/28805817f2f438a1626c33c38265ab3ec2511c9f. The solution was to use, e.g., **`SetCurrentDirectoryA` function instead of `SetCurrentDirectory`**. Unfortunately, it is not documented in the [official docs](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setcurrentdirectory) that `SetCurrentDirectory` function is actually not present in `Kernel32.dll`. But at least we have that working now. The latest test on Windows in `Akirathan/chdir-native` repo succeeds - https://github.com/Akirathan/chdir-native/actions/runs/14218672786/job/39841141038

<blockquote><img src="https://opengraph.githubassets.com/03d528fa1e4147cc73f0b5323058d17028cd09c66e2763d72271b0d06f836bd6/Akirathan/chdir-native" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/Akirathan/chdir-native">Use proper FS separator char · Akirathan/chdir-native@ac45b63</a></strong></div><div>Does java.io in NI respect cwd? Contribute to Akirathan/chdir-native development by creating an account on GitHub.</div></blockquote>

Commit:
FIx implementation of WindowsWorkingDirectory.

Inspiration from https://github.com/Akirathan/chdir-native/commit/28805817f2f438a1626c33c38265ab3ec2511c9f

Commit:
Use File.separator instead of hard-coded one

Commit:
Use the same native impl on Linux and Mac

Commit:
Fix compilation error in configuration.scala - unused import

Commit:
Make building change-directory test native image verbose

Commit:
Make building change-directory test native image even more verbose

Commit:
WorkingDirectory is real singleton

Commit:
NativeImage build is not verbose

Commit:
PosixWorkingDirectory should be enabled on MacOS

Commit:
String does not need to end with 0

Commit:
Do not invoke pwd in PosixWorkingDirectory

Commit:
Test consistency between getcwd and pwd on Unix

Commit:
Test consistency between getcwd and pwd on Unix

Commit:
Replace Linux with Unix in PosixWorkingDirectory

Commit:
[WIP] More logging

Commit:
Revert "[WIP] More logging"

This reverts commit bc6d63a793cf713a00e013f22cfeb469bf5219aa.

Commit:
Add test for symlinks

Commit:
Ensure that the expected path in test does not contain symlinks

Akirathan commented:
Recent failure in `change-directory` native image tests on Mac at https://github.com/enso-org/enso/actions/runs/14240156505/job/39908017390?pr=12618#step:10:2508 were caused by the fact that `getcwd` reports **real** path, without any symlinks, and in the test, I was expecting path with symlink parts. It should be fixed by https://github.com/enso-org/enso/pull/12618/commits/40f357b7c943e4673b9c52863079c08bd2d04656.

Moreover, added an explicit test for symlink in https://github.com/enso-org/enso/pull/12618/commits/3e5f54704389290dff974dc0a66fffb507f70469

<blockquote><img src="https://opengraph.githubassets.com/0787af4c9dbc05a55a6316f024b077e97bb381555c39f2eb6ab4d0914bd70329/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Set cwd before executing a project · enso-org/enso@5b37f96</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Set cwd before executing a project · enso-org/enso@5b37f96</div></blockquote>

hubertp commented on lib/java/change-directory/src/test/java/org/enso/change/directory/TestRunner.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  private static void printFailures(List<Failure> failures) {
+    for (var failure : failures) {
```
Shouldn't this print to `System.err`?

Akirathan replied:
For some reason, `sbt` slurps whatever gets to `System.err` from a subprocess and does not show it. Hence printing to the `System.out`.

Commit:
Ensure that the expected path in test does not contain symlinks

Akirathan commented:
# What code in JDK causes `java.io.File` to ignore further wdir changes?

In short, `System.getProperty("user.dir")`. In other words, once `System.getProperty("user.dir")` is called, all further changes to the process working directory via native code are ignored by the `java.io` file system. Note that `user.dir` system prop read happens in the class initializer of `java.io.UnixFileSystem` (which is called from `java.io.File` class initializer). To prove this claim, apply this patch:
```diff
diff --git a/engine/runner/src/main/java/org/enso/runner/Main.java b/engine/runner/src/main/java/org/enso/runner/Main.java
index e8c6a234c1..ba6a1ffe9d 100644
--- a/engine/runner/src/main/java/org/enso/runner/Main.java
+++ b/engine/runner/src/main/java/org/enso/runner/Main.java
@@ -44,11 +44,13 @@ import org.enso.polyglot.PolyglotContext;
 import org.enso.polyglot.debugger.DebuggerSessionManagerEndpoint;
 import org.enso.profiling.sampler.NoopSampler;
 import org.enso.profiling.sampler.OutputStreamSampler;
+import org.enso.change.directory.WorkingDirectories;
 import org.enso.runner.common.LanguageServerApi;
 import org.enso.runner.common.ProfilingConfig;
 import org.enso.runner.common.WrongOption;
 import org.enso.version.BuildVersion;
 import org.enso.version.VersionDescription;
+import org.graalvm.nativeimage.ImageInfo;
 import org.graalvm.polyglot.PolyglotException;
 import org.graalvm.polyglot.PolyglotException.StackFrame;
 import org.graalvm.polyglot.SourceSection;
@@ -1343,6 +1345,8 @@ public class Main {
 
   private void launch(String[] args) throws IOException, InterruptedException, URISyntaxException {
     var line = preprocessArguments(args);
+    var curDir = System.getProperty("user.dir");
+    chdir();
 
     var logMasking = new boolean[1];
     var logLevel = setupLogging(line, logMasking);
@@ -1555,4 +1611,16 @@ public class Main {
     }
     return true;
   }
+
+  private static void chdir() {
+    var nativeApi = WorkingDirectories.getCurrent();
+    var dirChanged = nativeApi.changeWorkingDir("/home/pavel/Documents");
+    if (dirChanged) {
+      System.out.println("Changed working dir to ~/Documents");
+    } else {
+      System.out.println("Failed to change wdir to ~/Documents");
+    }
+    var f = new File("tmp.txt");
+    System.out.println("new File(\"tmp.txt\").getAbsolutePath() = " + f.getAbsolutePath());
+  }
 }
```
And run native image with, e.g., `enso --run ~/tmp/Proj`:
```
Changed working dir to ~/Documents
new File("tmp.txt").getAbsolutePath() = /home/pavel/dev/enso/tmp.txt
```
In the above output, we can see that `chdir("/home/pavel/Documents")` sys call succeeded, but the `java.io` file system ignores it.

-----------------------------------------------------------------

But if `System.getProperty("user.dir")` is removed, then it works as expected:
```
Changed working dir to ~/Documents
new File("tmp.txt").getAbsolutePath() = /home/pavel/Documents/tmp.txt
```

Commit:
Implement WorkingDirectory.isDirectory

Commit:
Include change-directory module in engine-runner

Commit:
Try changing working directory as soon as possible in the launcher

Commit:
Revert "Implement WorkingDirectory.isDirectory"

This reverts commit f212fbf7c51286bf4e380923ba6d994ef62761c6.

Akirathan commented:
First positive result is finally here - with https://github.com/enso-org/enso/pull/12618/commits/4f00f362363c98fd6619067145c646d650df6203, I am able to run `enso --run ~/tmp/Proj/src/Main.enso` with output:
```
+ env JAVA_OPTS= -ea /home/pavel/dev/enso/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --log-level info --run /home/pavel/tmp/Proj/src/Main.enso
[INFO] [2025-04-04T17:25:58+02:00] [org.enso.interpreter.runtime.DefaultPackageRepository] Found library Standard.Base @ 0.0.0-dev is  AOT ready at [***/0.0.0-dev].
=== Enso File ===
path: /home/pavel/tmp/MY_FILE.txt
exists: False
=== Java File ===
path: /home/pavel/tmp/MY_FILE.txt
exists: False
```

Current wdir is `~/dev/enso` and the process changes wdir to `~/tmp` which is a parent directory of the project root `~/tmp/Proj`. And there is no warning printed.

Commit:
Cleanup after TestChangeDirectory

Akirathan commented:
Tests are green. The only failure is in Path operations in Base_Tests in Standard Library Tests in https://github.com/enso-org/enso/actions/runs/14268803065/job/39997025728?pr=12618#step:10:1403 which is expected, because chdir does not work in JVM mode. I will try to enforce running that test only in AOT mode.

<blockquote><img src="https://opengraph.githubassets.com/dcbd754784f298f820fa597c22ba15a6f32bb26a26dcd40087ee9866d0be7ba9/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Set cwd before executing a project · enso-org/enso@4b3787c</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Set cwd before executing a project · enso-org/enso@4b3787c</div></blockquote>

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -1459,6 +1466,80 @@ final CommandLine preprocessArguments(String... args) {
+  private void maybeChangeWorkingDirToProjectRoot(String fileToRun) {
+    assert fileToRun != null;
+    if (!ImageInfo.inImageRuntimeCode()) {
+      logger.warn("Changing working directory only supported in native image");
```
- I assume there is going to be the "other warning" printed later
- No need to warn, if the current directory is correct
- Change to `logger.trace`, maybe?

Akirathan replied:
Good point. Removed in 7ef9913c2a36e213028b1eb3d53f27b2a814718d. No point in changing this to `logger.trace` as the logging is not initialized at this point, and the trace level would be discarded anyway.

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -1459,6 +1466,80 @@ final CommandLine preprocessArguments(String... args) {
+  private static String findProjectRoot(String path) {
+    var nativeApi = WorkingDirectories.getCurrent();
+    // Cannot use File.separator char here
+    var separatorChar =
```
Unused variable.

Akirathan replied:
Removed in 06258375af8e3538cf151b116351b2af0872448e

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -1459,6 +1466,80 @@ final CommandLine preprocessArguments(String... args) {
+  private static String parentFile(String path) {
+    // Cannot use File.separator char here
+    var separatorChar =
+        switch (Platform.getOperatingSystem()) {
```
Move this code to `Platform.separatorChar()`.

Akirathan replied:
Done in 1b00a68440815ba9c1afe70dd4b1c0fd57f70ff0

JaroslavTulach commented on lib/java/change-directory/src/main/java/org/enso/change/directory/PosixWorkingDirectory.java:
```diff
@@ -0,0 +1,116 @@
+import org.slf4j.LoggerFactory;
+
+@CContext(PosixWorkingDirectory.Directives.class)
+public final class PosixWorkingDirectory implements WorkingDirectory {
```
This class shouldn't be exported from the `change-directory` module:
```suggestion
final class PosixWorkingDirectory implements WorkingDirectory {
```

JaroslavTulach commented on lib/java/change-directory/src/main/java/org/enso/change/directory/WindowsWorkingDirectory.java:
```diff
@@ -0,0 +1,107 @@
+import org.slf4j.LoggerFactory;
+
+@CContext(WindowsWorkingDirectory.Directives.class)
+public final class WindowsWorkingDirectory implements WorkingDirectory {
```
```suggestion
final class WindowsWorkingDirectory implements WorkingDirectory {
```

JaroslavTulach commented on lib/java/change-directory/src/main/java/org/enso/change/directory/WindowsWorkingDirectory.java:
```diff
@@ -0,0 +1,107 @@
+
+  @Override
+  public boolean changeWorkingDir(String path) {
+    try (var cPath = CTypeConversion.toCString(path + "\0")) {
```
```suggestion
    try (var cPath = CTypeConversion.toCString(path)) {
```

JaroslavTulach commented on lib/java/change-directory/src/main/java/org/enso/change/directory/WorkingDirectories.java:
```diff
@@ -0,0 +1,17 @@
+        case WINDOWS -> WindowsWorkingDirectory.INSTANCE;
+      };
+
+  public static WorkingDirectory getCurrent() {
```
To minimize the conceptual surface:
- move this method to `WorkingDirectory`
- make this class package private
- or rather merge this into `WorkingDirectory` abstract class.

Akirathan replied:
Done in c01d17d35f59f9253d6c063a70253974e856381c

JaroslavTulach commented on lib/java/change-directory/src/main/java/org/enso/change/directory/WorkingDirectory.java:
```diff
@@ -0,0 +1,9 @@
+package org.enso.change.directory;
+
+public interface WorkingDirectory {
```
You don't want 3rd parties to implement this interface:
```suggestion
public abstract class WorkingDirectory {
    WorkingDirectory() {}
```

JaroslavTulach commented on lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/DesktopEnvironment.java:
```diff
@@ -0,0 +1,39 @@
+
+import org.enso.common.Platform;
+
+public final class DesktopEnvironment {
```
To minimize conceptual surface, make this package private:
```suggestion
final class DesktopEnvironment {
```
and move the static into `Directories.getCurrent()` and `TrashBin.getCurrrent()` to mimic the API of `WorkingDirectory`

JaroslavTulach commented on lib/java/desktop-environment/src/test/java/org/enso/desktopenvironment/DirectoriesTest.java:
```diff
@@ -2,12 +2,13 @@
 public class DirectoriesTest {
 
-  private static final Directories directories = Platform.getOperatingSystem().getDirectories();
+  private static final Directories directories = DesktopEnvironment.getDirectories();
```
This would become:
```suggestion
  private static final Directories directories = Directories.getCurrent();
```

JaroslavTulach commented on lib/java/desktop-environment/src/test/java/org/enso/desktopenvironment/TrashBinTest.java:
```diff
@@ -3,14 +3,15 @@
 public class TrashBinTest {
 
-  private static final TrashBin TRASH_BIN = Platform.getOperatingSystem().getTrashBin();
+  private static final TrashBin TRASH_BIN = DesktopEnvironment.getTrashBin();
```
This would become:
```suggestion
  private static final TrashBin TRASH_BIN = TrashBin.getCurrent();
```

JaroslavTulach commented on lib/java/desktop-environment/src/test/java/org/enso/desktopenvironment/DirectoriesTest.java:
```diff
@@ -2,12 +2,13 @@
 import org.junit.Assert;
 import org.junit.Test;
 
 public class DirectoriesTest {
```
- shouldn't this test also run in _native image mode_?
- there doesn't seem to be the _NI testing_ infrastructure in `desktop-environment` project
- what about merging `desktop-environment` and `change-directory` into single `os-environment` project?
- then all the tests could run in _NI mode_

Akirathan replied:
Merged to single project in 59f7328965af29aa66a5e386de7476411cf5edb1

Akirathan replied:
Ensured that all these tests run in NI mode:
- 8c8cc2b5931663e01a312ea84cc8f7eadfc9cd63
- c6d04942de6295607dd12afa59ed74d04fd154e3

JaroslavTulach commented on lib/java/change-directory/src/test/java/org/enso/change/directory/TestChangeDirectory.java:
```diff
@@ -0,0 +1,128 @@
+        inNativeImage);
+  }
+
+  private static void ensureOnUnix() {
```
Why don't we test on Windows?

Akirathan replied:
Only one test is ignored on Windows - [curDir_IsSameAsPwdOnUnix](https://github.com/enso-org/enso/blob/c01d17d35f59f9253d6c063a70253974e856381c/lib/java/change-directory/src/test/java/org/enso/change/directory/TestChangeDirectory.java#L43-L49) for obvious reason - `pwd` is not present on Windows. All the other tests run on all platforms.

JaroslavTulach reviewed (approved):
- few API notes left inside
- if this can CWD, then we are done, aren't we?
- consider creation of a merge `os-environment` module
- consider testing `TrashBin` in _NI mode_

Commit:
Merge branch 'develop' into wip/akirathan/12301-fix-cwd-before-executing-main

Commit:
Apply suggestions from code review

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>

Commit:
Fix compilation after review apply

Commit:
Try to enable symlink test on Windows.

Commit:
Remove unused var

Commit:
Move separatorChar to Platform

Commit:
Merge WorkingDirectories to WorkingDirectory abstract class

Commit:
fmt

Commit:
Fix compilation error

Commit:
Merge desktop-environment and change-directory to single os-environment project.

All tests will be in NI.

Commit:
Add transitive module dependencies to os-environment

Commit:
Fix scala compiler errors

Commit:
Specify the list of native image tests only once

Commit:
Run all tests in NI

Commit:
Run all TrashBinTests on the CI

Commit:
Keep DirectoriesTest.getDocuments ignored on Windows

Commit:
Do not run the relative path resolve test in JVM mode

Commit:
Do not print duplicated warning

JaroslavTulach commented on engine/common/src/main/java/org/enso/common/Platform.java:
```diff
@@ -48,6 +35,19 @@ public static Platform getOperatingSystem() {
+   *
+   * @return the file separator character for the current operating system.
+   */
+  public static char separatorChar() {
```
- nice to see `separatorChar` method
- encapsulates the logic nicely
- when I was dreaming about it
- I was dreaming about **instance** method, not `static` one...
